pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java

```diff
@@ -510,25 +524,120 @@ private CompletableFuture<Integer> closeServiceUnit(String serviceUnit) {
+                return;
+            }
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+            ServiceUnitStateData next = new ServiceUnitStateData(Owned, data.broker());
+            for (NamespaceBundle sBundle : splitBundles.getRight()) {
+                futures.add(pubAsync(sBundle.toString(), next).thenAccept(__ -> {}));
+            }
+            NamespaceName nsname = bundle.getNamespaceObject();
+            FutureUtil.waitForAll(futures).thenRun(() ->
+                    namespaceService.updateNamespaceBundles(nsname, splitBundles.getLeft()).thenCompose(__ ->
+                            namespaceService.updateNamespaceBundlesForPolicies(nsname, splitBundles.getLeft()))
+                    .thenRun(() -> {
+                        bundleFactory.invalidateBundleCache(bundle.getNamespaceObject());
+                        updateFuture.complete(splitBundles.getRight());
+                    }).exceptionally(ex -> {
+                        log.warn("Failed to update namespace policies [{}], NamespaceBundle: {} due to {}",
+                                nsname, bundle.getBundleRange(), ex.getMessage());
+                        updateFuture.completeExceptionally(ex);
+                        return null;
+                    })).exceptionally(e -> {
+                        updateFuture.completeExceptionally(e);
+                        return null;
+                    });
+        }).exceptionally(e -> {
+            updateFuture.completeExceptionally(e);
+            return null;
+        });
+
+        updateFuture.thenAccept(r -> {
+            // Free the old bundle
+            tombstoneAsync(serviceUnit).thenRun(() -> {
+                // Update bundled_topic cache for load-report-generation
+                pulsar.getBrokerService().refreshTopicToStatsMaps(bundle);
+                // TODO: Update the load data immediately if needed.
+                completionFuture.complete(null);
+                double splitBundleTime = TimeUnit.NANOSECONDS.toMillis((System.nanoTime() - startTime));
+                log.info("Successfully split {} parent namespace-bundle to {} in {} ms", serviceUnit, r,
+                        splitBundleTime);
+            }).exceptionally(e -> {
+                double splitBundleTime = TimeUnit.NANOSECONDS.toMillis((System.nanoTime() - startTime));
+                String msg = format("Failed to free bundle %s in %s ms, under namespace [%s] with error %s",
+                        bundle.getNamespaceObject().toString(), splitBundleTime, bundle, e.getMessage());
+                completionFuture.completeExceptionally(new BrokerServiceException.ServiceUnitNotReadyException(msg));
+                return null;
+            });
+        }).exceptionally(ex -> {
+            // Retry several times on BadVersion
+            if ((ex.getCause() instanceof MetadataStoreException.BadVersionException)
```

Comment:
Where do we get this `MetadataStoreException.BadVersionException` from?  it seems like we are not accessing the metadata store in this block.

Reply:
The `namespaceService.updateNamespaceBundles(nsname, targetNsBundle)` and `namespaceService.updateNamespaceBundlesForPolicies(nsname, targetNsBundle)` might throw the `MetadataStoreException.BadVersionException`.